consensus/hotstuff/eventloop/event_loop.go

```diff
@@ -302,20 +299,22 @@ func (el *EventLoop) OnTcConstructedFromTimeouts(tc *flow.TimeoutCertificate) {
 // OnPartialTcCreated created a hotstuff.PartialTcCreated payload and pushes it into partialTcCreated buffered channel for
 // further processing by EventHandler. Since we use buffered channel this function can block if buffer is full.
 func (el *EventLoop) OnPartialTcCreated(view uint64, newestQC *flow.QuorumCertificate, lastViewTC *flow.TimeoutCertificate) {
-
+	received := time.Now()
 	event := &hotstuff.PartialTcCreated{
 		View:       view,
 		NewestQC:   newestQC,
 		LastViewTC: lastViewTC,
 	}
 
-	// TODO(active-pacemaker): maybe measure how long it takes for event to be processed
-
 	select {
 	case el.partialTcCreated <- event:
 	case <-el.ComponentManager.ShutdownSignal():
 		return
 	}
+
+	// the wait duration is measured as how long it takes from an event being
+	// received to event handler commencing the processing of the event
+	el.metrics.HotStuffWaitDuration(time.Since(received), metrics.HotstuffEventTypeOnPartialTc)
```

Comment:
[`e.partialTcCreated` has a channel size of 10](https://github.com/onflow/flow-go/blob/afdc0a830c6f61ff8ff40647d7a70b3c14982895/consensus/hotstuff/eventloop/event_loop.go#L51). So it's not measuring how long it takes for the Event Handler to start processing the event, it's measuring how long until the event is enqueued. 

This function is also blocking when the channel is full, which might block the `TimeoutProcessor` [here](https://github.com/onflow/flow-go/blob/94884ef4193fb6866060564a0c65d61e2ca2658f/consensus/hotstuff/timeoutcollector/timeout_processor.go#L142). 

Suppose we have 10 partial TC events queued, for views 10-20. Is it ever desirable to process the partial TC events for the views before 20? I don't think it is. So I'm wondering if we should use a pattern similar to how we process QCs and TCs here, where we store the latest partial TC by view in a `PartialTCTracker` and notify when it changes, rather than using a channel.

Reply:
The problem of measuring time-to-queueing rather than time-to-start-processing is similar for the proposals queue. The difference there could be large, since the queue has a max size of 1000 proposals. To measure how much time elapses between the EventLoop accepting the proposal and commencing processing, we could attach a timestamp to the proposal when enqueueing it in the channel, then record the difference between then and now when dequeueing the proposal in the event loop.

Reply:
https://github.com/onflow/flow-go/pull/3623/commits/a30e64688e838f0e4651a448e9d2d194085670a9